design/design-doc-package-refactor.md

```diff
@@ -0,0 +1,690 @@
+allows the controller to watch resources in its own namespace.
+
+The advantage of a `PackageInstall` is that is theoretically allows for multiple
+versions of the same package to be installed in a cluster at the same time
+because its controller is only watching for objects in its own namespace.
+However, because CRDs are cluster-scoped objects, there cannot be conflicting
+CRDs installed by two different versions of a `PackageInstall`. So while the
+goal of the `PackageInstall` is to enable multiple versions of a package to
+exist simultaneously in a cluster, the limitations of CRDs makes it less
+effective than desired in practice.
+
+## Goals
+
+The current package infrastructure, though well thought out, has become somewhat
+convoluted and redundant with the introduction of
+[composition](https://github.com/crossplane/crossplane/blob/master/design/design-doc-composition.md)
+into the Crossplane ecosystem.
+
+Composition solves the following goals originally intended to be addressed by a
+`PackageInstall` and [template
+stacks](https://github.com/crossplane/crossplane/blob/master/design/one-pager-template-stacks.md):
+
+- Ability to publish infrastructure abstractions to specific namespaces.
+  - The `PackageInstall` allowed packages to install a namespace-scoped CRD and
+    a controller that only watched for resources in its namespace. Composition
+    enables this by creating a namespace-scoped CRD using an
+    `InfrastructurePublication` and restricting the namespace in which it can be
+    provisioned using RBAC.
+- Ability to create new packages without writing any code.
+  - The templating controller enabled this by allowing a user to create a
+    `behavior.yaml` file and automatically using the templating controller image
+    (configured with the `behavior.yaml`) when one was not supplied in the
+    package. Composition enables this without having to install a new
+    controller. The composition controllers run as part of core Crossplane and
+    dynamically reconcile new types that are created in response to the creation
+    of an `InfrastructureDefinition` / `InfrastructurePublication`.
+
+Because composition serves the same purposes as these existing packaging
+primitives, the current packaging system can be narrowed in scope while also
+supporting a similar set of use-cases. The immediate goals of this refactor are:
+
+- Deprecation and subsequent removal of the `PackageInstall` and
+  `StackDefinition` types.
+- Standardization on a single `Package` type as a unit of installation.
+- Support for upgrading a `Package` that is installed and running in a cluster.
+- Support for native installation of composition types, including
+  `InfrastructureDefinition`, `InfrastructurePublication`, and `Composition`.
+
+The refactor will involve abandoning the following goals of the current model:
+
+- Supporting running multiple versions of the same package in a cluster at the
```

Comment:
Not having multiple versions of the same package installed in the same cluster definitely simplifies some things and I think that's a good idea.  The real draw for that in the past was to be able to create instances of the package's CRDs within more than 1 namespace, which isn't a good reason to take on the burden of managing multiple installations. I don't think people really wanted multiple versions of the package as much as they wanted usage from multiple namespaces.

I quite like the composition approach where 1 `InfrastructureDefinition` is installed in the cluster, then can be published to multiple namespaces with multiple `InfrastructurePublication`.  No complicated package conflicts, ref counting, etc.  It's much cleaner, and we still get the real benefit that we wanted all along. 👍